cmip6_downscaling/methods/bcsd.py

```diff
@@ -0,0 +1,249 @@
+        # the coarsened obs back to its original resolution and write it out (will be used in postprocess)
+        spatial_anomolies = get_spatial_anomolies(coarse_obs, obs_ds, variable, connection_string)
+        write_dataset(spatial_anomolies, spatial_anomolies_path, chunks_dims=('month',))
+
+    return coarse_obs_path, spatial_anomolies_path
+
+
+def prep_bcsd_inputs(
+    coarse_obs_path: str,
+    gcm: str,
+    scenario: str,
+    obs_id: str,
+    train_period_start: str,
+    train_period_end: str,
+    predict_period_start: str,
+    predict_period_end: str,
+    variable: str,
+) -> tuple[str, str, str]:
+    """Prepare the inputs to be fed into the training and fitting of the model. Largely
+    this converts any datasets chunked in time into ones chunked in space since
+    the models are pointwise and need the full timeseries to be performant.
+
+    Parameters
+    ----------
+    coarse_obs_path : str
+        Path to read coarsened obs zarr dataset (e.g. 'az://bucket/out.zarr')
+    gcm : str
+        Name of GCM
+    obs_id : str
+        Name of obs dataset (currently only supports 'ERA5')
+    train_period_start : str
+        Date for training period start (e.g. '1985')
+    train_period_end : str
+        Date for training period end (e.g. '2015')
+    predict_period_start : str
+        Date for prediction period start (e.g. '2070')
+    predict_period_end : str
+        Date for prediction period end (e.g. '2099')
+    variable : str
+        Variable of interest in CMIP conventions (e.g. 'tasmax')
+
+    Returns
+    -------
+    y_rechunked_path, X_train_rechunked_path, X_predict_rechunked_path : str, str, str
+        Paths to to write y, X_train, and X_predict ready for eventual use by model
+    """
+    # load in coarse obs as xarray ds
+    [y] = load_paths([coarse_obs_path])
+    # TODO: add xarray_schema test here for chunking - if the chunking fails then try rechunking
+    # if passes skip
+    y_rechunked, y_rechunked_path = rechunk_zarr_array(
+        y,
+        chunk_dims=('lat', 'lon'),
+        variable=variable,
+        connection_string=connection_string,
+        max_mem='1GB',
+    )
+
+    X_train = load_cmip(source_ids=[gcm], return_type='xr').sel(
+        time=slice(train_period_start, train_period_end)
+    )
+    X_train['time'] = y.time.values
+
+    X_train_rechunked, X_train_rechunked_path = rechunk_zarr_array(
+        X_train,
+        variable=variable,
+        chunk_dims=('lat', 'lon'),
+        connection_string=connection_string,
+        max_mem='1GB',
+    )
+
+    X_predict = load_cmip(
+        source_ids=[gcm], activity_ids=['ScenarioMIP'], experiment_ids=[scenario], return_type='xr'
+    ).sel(time=slice(predict_period_start, predict_period_end))
+    assert len(X_predict.lat) == len(
+        X_train.lat
+    ), "Uh oh! Your prediction length is different from your training and so the chunk sizes wont match and xarray will complain!"
```

Comment:
And, if this check is needed, lets raise a `ValueError` here instead of using an assert statement.

Reply:
this was a work-around but i implemented a stronger fix. the pointwise models need the spatial chunks to be the same between fit and predict steps. since we were algorithmically determining a 100mb chunk size (as opposed to specifying it), if we fed a different length of temporal record to predict, it might result in a different spatial chunk size. now i allow for either allowing the chunk size calculation or for specfying it. but it's good to keep in mind that this could cause a performance issue if we predict on much longer timescale than we train, since the spatial chunksize  (and thus the entire chunksize since there is only ever one chunk in time for pointwise models) will be determined by the train period, and it could result in larger chunks. for example, if chunk size of 30 years in train were 100mb, using the same spatial chunk but extending to a 150 year predict window would result in 500mb chunks.